### ISSUE TITLE
Failed to open X display when running in docker

### ISSUE THREAD

**Issue by ionorb** [maptiler/tileserver-gl#1285]
Hello there. I'm using tileserver-gl v4.11.1 and I'm getting the classic:

```
terminate called after throwing an instance of 'std::runtime_error'
   what(): Failed to open X display.
```

Normally this happens when the x display server is not running, however I'm running with docker... Are there any suggestions on how I can investigate this issue? I'm not sure where to start. This issue happened a few times on my local machine, but restarting would do the trick, but now it is hosted on a remote server and it is happening every time.

**Comment by nathanpackard** [comment]
I'm having the same problem.

**Comment by acalcutt** [comment]
Does it work either of you with previous versions? what is the last version that worked for you.

**Comment by nathanpackard** [comment]
I am actually running: Tileserver-gl v4.4.10, and get this error.

I am new to the project I'm working on so can't comment on the last version that worked.

**Comment by acalcutt** [comment]
The docker image uses [xvfb](https://github.com/maptiler/tileserver-gl/blob/master/docker-entrypoint.sh#L6) to provide X display. So the only thing i could think of is your cpu doesn't meet the  requirements for it to emulate open-gl. What cpu and os are you running on?

I ran into something like this running directly on windows 2022 when i ran on a vitual server, since it didn't support opengl.

the crash is likely happening when you visit the index page, where it needs to render thumbnails, or loading a rendered tiles. since this is when maplibre-native needs X display to render.

**Comment by nathanpackard** [comment]
Yeah it sounds like a resource issue. however, I have lots of resources. Here is my system summary:
![image](https://github.com/maptiler/tileserver-gl/assets/6738377/a0e282df-9856-44a7-9e54-8044da84d122)

Also, my docker-compose.yml file has a lot of resources assigned:
    mem_limit: 24G
    cpus: '8.0'
    deploy:
      resources:
        limits:
          cpus: '8.0'
          memory: 24G
        reservations:
          cpus: '8.0'
          memory: 24G

**Comment by acalcutt** [comment]
Are you able to test swapping these two packages around and see if it makes a difference
https://github.com/maptiler/tileserver-gl/blob/master/src/serve_rendered.js#L5-L11

**Comment by asukachiharu** [comment]
> We have found that when using xvfb in maplibre-native ci workflows, it does fail with that error sometimes. it seemed to be a known xvfb issue

I found that as long as there is a style.json file in the folder, starting the Docker container will result in this error. However, this issue almost never occurs on Windows.

**Comment by asukachiharu** [comment]
> > We have found that when using xvfb in maplibre-native ci workflows, it does fail with that error sometimes. it seemed to be a known xvfb issue
> 
> I found that as long as there is a style.json file in the folder, starting the Docker container will result in this error. However, this issue almost never occurs on Windows.

Specifically, this error does not appear locally. However, when previewing the raster, an error occurs: [error: failed to parse json: the document is empty. at offset 0] /GET xxxxx/xx512/0/0/0.png 500.

**Comment by docuracy** [comment]
I guess this is the same problem:

```text
2024-12-31T22:47:14.607Z | [CI] Failed to open X display, retrying...
... (20 of these retry notifications in total) ...
2024-12-31T22:47:24.115Z | [CI] Failed to open X display, retrying...
2024-12-31T22:47:24.615Z | terminate called after throwing an instance of 'std::runtime_error'
2024-12-31T22:47:24.615Z |   what():  Failed to open X display.
```

It's triggered by _any_ call for a static map, for example:

```mathematica
curl -I "http://localhost:30080/styles/elevation/static/9.051,48.228,10/1x1.png"
```

I'm running the [Docker v5.0.0 image](https://hub.docker.com/layers/maptiler/tileserver-gl/v5.0.0/images/sha256-edd142ff22da38904af6691e86611be5c8d48df649433fc04b86a5a63f22e74f) in Kubernetes (which includes `xvfb` , intended I understand to run as a daemon when required), with these resources:

```arduino
    requests:
      memory: "2Gi"
      cpu: "2"
    limits:
      memory: "4Gi"
      cpu: "4"
```

Any suggestions, please?

**Comment by mloskot** [comment]
I'm experiencing the same issue, as in @docuracy  https://github.com/maptiler/tileserver-gl/issues/1285#issuecomment-2566771907, when running latest [maptiler/tileserver-gl:v5.1.3](https://hub.docker.com/layers/maptiler/tileserver-gl/v5.1.3/images/sha256-3cf119ee23bf313a42805aa7d5d86f40d74574d10d6464242f8a87eae42c8865) from container on AKS cluster.
Thumbnails of previews do not load and attempt to directly access thumbnail URL crashes TileServer-GL, and terminates the container.

**Comment by okimiko** [comment]
As of @mloskot's issue is caused by an entrypoint overwrite (`command` == `ENTRYPOINT`, `args` == `CMD`, see https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#define-a-command-and-arguments-when-you-create-a-pod), it may be helpful to check the startup config of the others, too.

@mloskot I like the approach of configuration by environment variables, but I'm not sure if this always fits every need. I mount a preconfigured file in the container. May be this is an option for you, too (see https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/).

**Comment by okimiko** [comment]
@mloskot AFAIK: None and in my opinion that is fine: I'm not a fan of long process calls. For simple needs it may help to have some more command line options, but for that I would preferer environment variables. But (as said before) that is just my opinion.

PS: I just checked the source; next to UV_THREADPOOL_SIZE, there is already something for PORT/BIND and a development checks, I thinks it should be not that hard to extend this for non-list options.

PPS: I think we are a bit offtopic, this may be something for a discussion (or a PR ;-))

**Comment by mloskot** [comment]
@okimiko 
>  None and in my opinion that is fine: I'm not a fan of long process calls.

You seem to be misunderstanding my suggestion. The fact command line options cannot be controlled in `config.json` leads users to tweaking them via container command line and THAT leads to the run-time problems like the one discussed in this issue 

If users could stick to [default entrypoint of the official container image](https://github.com/maptiler/tileserver-gl/blob/78a88c1857c8c66943a09e310df6ddc731719cfb/docker-entrypoint.sh) in **every** scenario, they would have not experienced craches. I'm happy to learn I'm wrong about this.

Having said that, with respect, but _"how can we let users configure containers with all aspects of TileServer-GL without touching command line arguments"_ is absolutely in-topic here as addressing that question is part of solution of this issue here.

**Comment by okimiko** [comment]
Yes, you are right then, I misunderstood your suggestion. But then I have an even more opposit opinion: You already can append every command line option (using `args` or `CMD` or `command`) if you do not overwrite the default entrypoint (like in every other image, too).

**Comment by mloskot** [comment]
@okimiko 
> You already can append every command line option (using `args` or `CMD` or `command`) if you do not overwrite the default entrypoint 

You cannot use the `command` without overwriting the default entry point.

From https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/
> The `command` field corresponds to `ENTRYPOINT`, and the `args` field corresponds to `CMD` in some container runtimes.

If users need to specify `--public_url`, they will apply the obvious canonical manifest:

```yaml
command: [ "node" ]
args: [ "/usr/src/app", "--public_url", "..." ]
```

then, TileServer-GL will start crashing at them!

If you can't see there is room for improvement around the configuration aspect here to help users avoid shooting their feet, then I can't help, but I feel like we are running in circles. So, I will to stop there.

**Comment by okimiko** [comment]
Then stop using `command`. My point is, that I do not understand what prevents you from just using this:

```yaml
args: [ "--public_url", "..." ]
```
